Network e Repository(Paging) layers implementados.

LucasCabralDevv opened the pull request:
Foi criado: 

a camada de network em um modulo usando koin.

a classe de paginação (usando a biblioteca paging3), que esta sendo utilizada no viewModel dentro da função getListData. A activity através desta função observará as mudanças de paginação e notificará o Adapter.

OBS: Futuramente penso em fazer algumas modificações para deixar o código escalável e melhor testável. A classe criada para fazer a paginação não trabalha como um repositório em si. Como estou usando o paging3 pela primeira vez, resolvi fazer primeiro dessa maneira. Também penso em fazer cache das páginas usando room.
Após fazer essa modificações, consigo trabalhar melhor com o koin e criar os modulos para o repository e viewModel.

Link trello:
https://trello.com/c/XrDUhcKv/49-projeto-final-marvel-cria%C3%A7%C3%A3o-da-camada-de-repository-do-projeto

Commit:
Network e Repository(Paging) layers implementados.

Lhmsouza commented on .idea/discord.xml:
```diff
@@ -0,0 +1,6 @@
+  <component name="DiscordProjectSettings">
+    <option name="show" value="PROJECT" />
+  </component>
+</project>
```
O que faz esse arquivo aqui?

LucasCabralDevv replied:
Tenho quase certeza que é o plugin do discord que eu instalei. Sempre quando entro no projeto, fica amostra no perfil do discord (como por exemplo o spotify).

Lhmsouza replied:
Que massa, imaginei que era mesmo. Essas configurações dentro do `.idea` daria pra deixar só local, mas isso dá pra ver como fazer numa outra hora.

Lhmsouza commented on app/build.gradle:
```diff
@@ -35,11 +36,35 @@ android {
-    implementation 'androidx.core:core-ktx:1.3.2'
-    implementation 'androidx.appcompat:appcompat:1.2.0'
+    implementation 'androidx.core:core-ktx:1.5.0'
+    implementation 'androidx.appcompat:appcompat:1.3.0'
```
Vi que usou versões com `$[...]_version` nas outras dependências. Será que teria como fazer pra estas daqui também?

LucasCabralDevv replied:
Posso sim.

Lhmsouza commented on app/build.gradle:
```diff
@@ -35,11 +36,35 @@ android {
+
+    // Lifecycle Dependencies
+    implementation 'androidx.lifecycle:lifecycle-extensions:2.2.0'
+    implementation 'androidx.lifecycle:lifecycle-viewmodel-ktx:2.3.1'
```
Mesma coisa do outro comentário nesse arquivo.

LucasCabralDevv replied:
Feito.

Lhmsouza commented on app/build.gradle:
```diff
@@ -35,11 +36,35 @@ android {
     androidTestImplementation 'androidx.test.espresso:espresso-core:3.3.0'
+    // Koin & Mock test Dependencies
+    testImplementation "org.koin:koin-test:$koin_version"
+    testImplementation 'com.nhaarman:mockito-kotlin:1.5.0'
```
Mesma coisa do outro comentário nesse arquivo.

LucasCabralDevv replied:
Feito.

Lhmsouza commented on app/src/main/java/com/lucascabral/marvelsuperheroes/helper/ConstantKeys.kt:
```diff
@@ -0,0 +1,4 @@
+package com.lucascabral.marvelsuperheroes.helper
+
+const val API_KEY: String = "426a8a3c32e2543316228982d5e3d552"
+const val PRIVATE_KEY: String = "3d286fee9404e7f0c565a623ccd54871b179a2b6"
```
Estas chaves poderiam estar no BuildConfig, mais ou menos como te mostrei naquele dia, dos [Build Config Fields](https://developer.android.com/studio/build/gradle-tips#share-custom-fields-and-resource-values-with-your-app-code). Recomendo porque, imaginando um caso mais real de uso de chaves de API, provavelmente não teríamos ela gravada no código; mas, se tivéssemos, seria mais interessante manter essa chave oculta dentro do ambiente de deploy, pra não precisar manter ela no código. Aí, no caso, entra naquele esquema de termos dois ou mais build types (`debug` e `release`), que não é exatamente o nosso caso aqui, mas fica a recomendação.

Ficariam até com o mesmo nome, só seriam acessadas por `BuildConfig.API_KEY` em vez de direto com `API_KEY`. Se não entender totalmente, tranquilo, depois podemos falar mais sobre.

LucasCabralDevv replied:
Feito.

Lhmsouza commented on app/src/main/java/com/lucascabral/marvelsuperheroes/presenter/MarvelApp.kt:
```diff
@@ -0,0 +1,19 @@
+package com.lucascabral.marvelsuperheroes.presenter
```
Algum motivo específico pra esse pacote se chamar `presenter`?

LucasCabralDevv replied:
Na arquitetura clean usamos presenter para a camada de apresentação (ui/adapter/), no caso penso em usar as duas arquiteturas no futuro clean e mvvm - se for interessante.

Lhmsouza commented on app/src/main/java/com/lucascabral/marvelsuperheroes/presenter/di/NetworkModule.kt:
```diff
@@ -0,0 +1,63 @@
+
+val networkModule = module {
+
+    single(named("BASE_URL")) { "http://gateway.marvel.com/v1/public/" }
```
Chegou a tentar usar o `BuildConfig` pra salvar essa String de URL? Comentei num outro ponto do código sobre os (Build Config Fields)[https://developer.android.com/studio/build/gradle-tips#share-custom-fields-and-resource-values-with-your-app-code], fica a recomendação de novo de tentar, sempre que possível, manter essas Strings importantes de uso geral ali. Mas dessa forma também ficou bem boa, só uma recomendação mesmo.

https://medium.com/@ramonrabello/best-practices-for-working-with-different-build-types-c5cd08b1adc6

Como comentei no outro ponto do código, não é o nosso caso aqui de ter build types e rotas diferentes, de `debug`, `release`, `homolog`, enfim. Então essas vantagens de aplicar a BASE_URL no BuildConfig acabam não sendo tão grandes pra nós. Mas recomendo mais por ser um conhecimento interessante e uma boa prática em projetos futuros. :wink:

LucasCabralDevv replied:
Entendido.

Lhmsouza commented on app/src/main/java/com/lucascabral/marvelsuperheroes/presenter/view/AllCharactersActivity.kt:
```diff
@@ -0,0 +1,26 @@
+    }
+
+    private fun initViewModel() {
+        val viewModel = ViewModelProvider(this).get(AllCharactersViewModel::class.java)
```
Tu poderia usar o Koin pra injetar essa `ViewModel` também, em vez de usar o `ViewModelProvider`. Se não souber como, dá uma pesquisada na [documentação do Koin](https://insert-koin.io/) sobre como injetar `ViewModels`.

LucasCabralDevv replied:
Feito.

Lhmsouza commented on app/src/main/java/com/lucascabral/marvelsuperheroes/presenter/viewmodel/AllCharactersViewModel.kt:
```diff
@@ -0,0 +1,26 @@
+     private val apiService: MarvelService by inject()
+
+     fun getListData(): Flow<PagingData<Character>> {
+          return Pager(config = PagingConfig(pageSize = 20, maxSize = 200),
```
Esses valores fixos aqui seria interessante criar variáveis constantes pra eles, pra evitar [magic numbers](https://stackoverflow.com/questions/47882/what-is-a-magic-number-and-why-is-it-bad). Poderiam ficar dentro de um `companion object` dentro dessa `ViewModel` mesmo.

LucasCabralDevv replied:
Feito.

Lhmsouza commented on build.gradle:
```diff
@@ -1,12 +1,19 @@
     }
     dependencies {
-        classpath "com.android.tools.build:gradle:4.2.0"
+        classpath "com.android.tools.build:gradle:4.2.1"
```
Será que seria possível usar um `$gradle_version`, por exemplo, pra manter o padrão?

LucasCabralDevv replied:
Feito.

Lhmsouza commented on app/src/main/java/com/lucascabral/marvelsuperheroes/presenter/di/NetworkModule.kt:
```diff
@@ -0,0 +1,63 @@
+        val logging = HttpLoggingInterceptor()
+        logging.level = HttpLoggingInterceptor.Level.BODY
+        logging
+    } */
```
Só mais como um comentário mesmo, pode deixar aqui por enquanto porque vai ser útil pra testar mais pra frente: normalmente não é muito recomendado deixar código comentado nos projetos. Cria muita dúvida de por que foi comentado, se é necessário, se pode ser necessário no futuro, acaba afetando a legibilidade.

Não precisa tirar agora, como falei, só fazendo esse comentário pra tu já ter uma ideia do porquê.

LucasCabralDevv replied:
Entendido.

Lhmsouza reviewed (changes requested):
No geral, muito bom, bem organizado e seguindo boas práticas de programação. Deixei alguns comentários e alguns links de recomendações, mas pra um primeiro PR tá super bacana. 😄

Commit:
Todas as dependências foram padronizadas.

Commit:
Keys foram passadas para o BuildConfig com type debug.

Commit:
Criado o modulo ViewModelModule.kt. ViewModel sendo injetado na activite.
Todos os testes de modulos do koin passaram.
Criado o companion object para evitar magic numbers.

Lhmsouza commented on app/src/main/java/com/lucascabral/marvelsuperheroes/presenter/di/ViewModelModule.kt:
```diff
@@ -0,0 +1,9 @@
+
+val viewModelModule = module {
+
+    single { AllCharactersViewModel() }
```
Aqui poderia usar `viewModel { AllCharactersViewModel() }`, mas acredito que com um `single` funcione também.

LucasCabralDevv replied:
Vou testar.
